### PR TITLE
DAOS-10227 vos: Fix a bug and add more bits to timestamp

### PR DESCRIPTION
**Pull request by jolivier23**
For RDB to perform 1000 txns/s, we need more bits to store
an epoch.   Use 42-bits of HLC instead of 32.  This gives us
roughly 116 years worth of epochs.

Also, don't update the epoch if it's lower than the stored
one.  This is a bug.

Signed-off-by: Jeff Olivier <jeffrey.v.olivier@intel.com>

### CODE REVIEW

File: src/vos/vos_internal.h [hunk 1]
```diff
@@ -365,64 +365,79 @@ extern uint64_t vos_evt_feats;
...
+/** Number of bits to use for timestamp (roughly 1/4 ms granularity) */
+#define VOS_AGG_NR_BITS	42
+/** HLC differentiation bits */
+#define VOS_AGG_NR_HLC_BITS	(64 - VOS_AGG_NR_BITS)
+/** Lower bits of HLC used for rounding */
+#define VOS_AGG_HLC_BITS	((1ULL << VOS_AGG_NR_HLC_BITS) - 1)
+/** Mask to check if epoch is HLC. */
+#define VOS_AGG_HLC_MASK	(VOS_AGG_HLC_BITS << (64 - VOS_AGG_NR_HLC_BITS))
+/** Mask of bits of epoch */
+#define VOS_AGG_EPOCH_MASK	(~VOS_AGG_HLC_MASK << (64 - VOS_AGG_NR_BITS))
```
Comment (NiuYawei) [inline src/vos/vos_internal.h:379]: Is it "~VOS_AGG_HLC_BITS"?

File: src/vos/vos_internal.h [hunk 2]
```diff
@@ -365,64 +365,79 @@ extern uint64_t vos_evt_feats;
...
-		*epoch += 0xffffffffULL;
+	if (!vos_feats_agg_time_get(*feats, &old_epoch))
+		old_epoch = 0;
+
+	if (epoch <= old_epoch) /** Only need to save newest */
+		return;
+
+	if (epoch & VOS_AGG_HLC_MASK) {
+		/** We only save the top bits so round up so the stored timestamp works */
+		epoch += VOS_AGG_HLC_BITS;
```
Comment (NiuYawei) [inline src/vos/vos_internal.h:430]: Why do we need to add the lower bits on updating? Shouldn't it being added on getting?

File: src/vos/vos_internal.h [hunk 3]
```diff
@@ -365,64 +365,79 @@ extern uint64_t vos_evt_feats;
...
-		epoch &= 0xffffffff00000000ULL;
-		/** Ensure the resulting epoch is not zero regardless, mostly for standalone */
-		epoch = (epoch >> VOS_TF_AGG_RSHIFT);
-		extra_flag = VOS_TF_AGG_HLC;
-	} else {
-		epoch &= 0x0ffffffffULL;
-		epoch = (epoch << VOS_TF_AGG_LSHIFT);
-	}
+	if (feats & VOS_TF_AGG_HLC)
+		*epoch = ((feats & VOS_AGG_TIME_MASK) << VOS_AGG_RSHIFT);
```
Comment (NiuYawei) [inline src/vos/vos_internal.h:408]: missed adding VOS_AGG_HLC_BITS?

File: src/vos/vos_internal.h (hunk 1 above)
Comment (liw) [reply src/vos/vos_internal.h:379]: Looks correct to me, though "(64 - VOS_AGG_NR_BITS)" could be replaced with "VOS_AGG_NR_HLC_BITS".

File: src/vos/vos_internal.h [hunk 4]
```diff
@@ -365,64 +365,79 @@ extern uint64_t vos_evt_feats;
...
 /** Start bit of stored timestamp */
 #define VOS_TF_AGG_BIT	60
-/** Right shift for 32-bits of epoch */
-#define VOS_TF_AGG_RSHIFT	(63 - VOS_TF_AGG_BIT)
-/** Left shift for 32-bits of epoch */
-#define VOS_TF_AGG_LSHIFT	(32 - VOS_TF_AGG_RSHIFT)
+/** Right shift for stored portion of epoch */
+#define VOS_AGG_RSHIFT	(63 - VOS_TF_AGG_BIT)
+/** Left shift for stored portion of epoch */
+#define VOS_AGG_LSHIFT	(VOS_AGG_NR_BITS - VOS_AGG_RSHIFT)
```
Comment (liw) [inline src/vos/vos_internal.h:385]: Should this be "(64 - VOS_AGG_NR_BITS - VOS_AGG_RSHIFT)", that is, 19?

    [63, 61]: flags // RSHIFT?
    [60, 19]: stored portion of epoch
    [18,  0]: something else // LSHIFT?

File: src/vos/vos_internal.h (hunk 2 above)
Comment (jolivier23) [reply src/vos/vos_internal.h:430]: I'm just rounding up to nearest 1/4 ms.    If I had done it this way to start with, it would be forward compatible.

File: src/vos/vos_internal.h (hunk 4 above)
Comment (jolivier23) [reply src/vos/vos_internal.h:385]: I need to add some tests for these corner cases.  I'll have a closer look.

File: src/vos/vos_internal.h (hunk 1 above)
Comment (jolivier23) [reply src/vos/vos_internal.h:379]: will fix

File: src/vos/vos_internal.h (hunk 2 above)
Comment (NiuYawei) [reply src/vos/vos_internal.h:430]: Ok, I see. Thanks.

**Review by NiuYawei — APPROVED** [review]

**Review by liw — APPROVED** [review]
Looks good to me AFAICS.

**Review by gnailzenh — APPROVED** [review]

**Review by Nasf-Fan — APPROVED** [review]
